codegen/src/main/ops/org/nd4j/codegen/ops/SDBaseOps.kt

```diff
@@ -0,0 +1,1806 @@
+        javaPackage = "org.nd4j.linalg.api.ops.impl.shape"
+        Input(NUMERIC, "df") {count = AtLeast(1); description = "" }
+        Arg(INT, "axis") { description = "" }
+        Output(NUMERIC, "output"){ description = "" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                @see #repeat(String, SDVariable, int)
+            """.trimIndent()
+        }
+    }
+
+    Op("replaceWhere") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.transforms.comparison"
+        javaOpClass = "CompareAndReplace"
+        legacy = true
+        Input(NUMERIC, "update") { description = "Source array" }
+        Input(NUMERIC, "from") { description = "Replacement values array (used conditionally). Must be same shape as 'update' array" }
+        Arg(CONDITION, "condition") { description = "Condition to check on update array elements" }
+        Output(NUMERIC, "output"){ description = "New array with values replaced where condition is satisfied" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Element-wise replace where condition:
+                out[i] = from[i] if condition(update[i]) is satisfied, or
+                out[i] = update[i] if condition(update[i]) is NOT satisfied
+            """.trimIndent()
+        }
+    }
+
+    Op("reshape") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.shape"
+        Input(NUMERIC, "x") { description = "Input variable" }
+        Input(NUMERIC, "shape") { description = "New shape for variable" }
+        Output(NUMERIC, "output"){ description = "Output variable" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Reshape the input variable to the specified (fixed) shape. The output variable will have the same values as the
+                input, but with the specified shape.
+                Note that prod(shape) must match length(input) == prod(input.shape)
+            """.trimIndent()
+        }
+    }
+
+    Op("reverse") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.transforms.custom"
+        Input(NUMERIC, "x") { description = "Input variable" }
+        Arg(INT, "dimensions") { count = AtLeast(1); description = "Input variable" }
+        Output(NUMERIC, "output"){ description = "Output variable" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Reverse the values of an array for the specified dimensions
+                If input is:
+                [ 1, 2, 3]
+                [ 4, 5, 6]
+                then
+                reverse(in, 0):
+                [3, 2, 1]
+                [6, 5, 4]
+                reverse(in, 0):
+                [4, 5, 6]
+                [1, 2 3]
+            """.trimIndent()
+        }
+    }
+
+    Op("reverseSequence") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.transforms.custom"
+        Input(NUMERIC, "x") { description = "Input variable" }
+        Input(NUMERIC, "seq_lengths") { description = "Length of the sequences" }
```

Comment:
seq_lengths should be integer

Reply:
fixed. (We still need to implement the new op constructors for these.)